internal/config/resource_base.go

```diff
@@ -6,19 +6,53 @@ import (
 
 	"github.com/hashicorp/terraform-plugin-framework/datasource"
 	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/tfsdk"
+	"github.com/hashicorp/terraform-plugin-framework/types"
 )
 
 const (
 	errorConfigureSummary = "Unexpected Resource Configure Type"
 	errorConfigure        = "expected *MongoDBClient, got: %T. Please report this issue to the provider developers"
 )
 
+type ProviderMeta struct {
+	ModuleName     types.String `tfsdk:"module_name"`
+	ModuleVersion  types.String `tfsdk:"module_version"`
+	UserAgentExtra types.Map    `tfsdk:"user_agent_extra"`
+}
+
+type ImplementedResource interface {
+	resource.ResourceWithImportState
+	// Additional methods such as upgrade state & plan modifier are optional
+	SetClient(*MongoDBClient)
+	GetName() string
+}
+
+func AnalyticsResourceFunc(iResource resource.Resource) func() resource.Resource {
+	commonResource, ok := iResource.(ImplementedResource)
+	if !ok {
+		panic("resource %T didn't comply with the ImplementedResource interface")
+	}
+	a := func() resource.Resource {
+		return analyticsResource(commonResource)
+	}
+	return a
+}
+
+func analyticsResource(iResource ImplementedResource) resource.Resource {
+	return &RSCommon{
+		ResourceName:        iResource.GetName(),
+		ImplementedResource: iResource,
+	}
+}
```

Comment:
```suggestion
func analyticsResource(iResource ImplementedResource) resource.Resource {
	return iResource
}
```
Might be missing something but I believe this function is redundant, `ImplementedResource` already has `resource.ResourceWithImportState` interface which is a `resource.Resource`.

Reply:
Good catch, but the function is actually necessary! While `ImplementedResource` does implement `resource.Resource`, we need to wrap it with `RSCommon` to intercept CRUD operations. 

If we returned `iResource` directly, the analytics tracking wouldn't work - `RSCommon` is what adds the provider_meta information to the context before calling the underlying resource methods. Without the wrapper:
- `Create/Read/Update/Delete` would execute directly on the resource without adding analytics to context
- The `UserAgentTransport` would receive empty context and no analytics would be tracked

Added text in d5452ab66ba4ad2e8deb2fe6836299030ba5ebd1

Reply:
I see. I am mainly confused because all our resource implementations are already defined with `RSCommon` ([example](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/internal/service/project/resource_project.go#L41)), so we are going from resource implementation which is RSCommon returned as `resource.Resource`, which then in `AnalyticsResourceFunc` gets wrapped in `RSCommon` again.

Reply:
Yes, it is a bit confusing but I don't know a better way.
The alternative would be to pass the CRUD operations to the RSCommon and then have that wrap those methods. But that requires changing every single resource.
This could work a bit safer if new methods are added

Reply:
I'm trying to see how we could change each resource:

```go
func Resource() resource.Resource {
	return &rs{
		RSCommon: config.RSCommon{
			ResourceName: resourceName,
		},
	}
}
```

But I'm not sure how to do this in a "clean" way.
What I like about the current implementation is that it keeps the existing resources untouched.
So whenever you add a resource you don't need to worry about the analytics part unless you use a new TPF method that hasn't been wrapped.